server/src/com/thoughtworks/go/server/messaging/plugin/StageStatusPluginNotifier.java

```diff
@@ -16,49 +16,258 @@
         data.put("stage-name", stage.getIdentifier().getStageName());
         data.put("stage-counter", stage.getIdentifier().getStageCounter());
-        data.put("stage-state", stage.getState());
-        data.put("stage-result", stage.getResult());
-        data.put("create-time", timestampToString(stage.getCreatedTime()));
-        data.put("last-transition-time", timestampToString(stage.getLastTransitionedTime()));
+        data.put("stage-state", stage.getState().toString());
+        data.put("stage-result", stage.getResult().toString());
+
+        Pipeline pipeline = pipelineService.fullPipelineById(stage.getPipelineId());
+        Map<String, Object> pipelineMap = createPipelineDataMap(pipeline);
+        data.put("pipeline", pipelineMap);
+
         return data;
     }
 
+    private Map<String, Object> createPipelineDataMap(Pipeline pipeline) {
+        Map<String, Object> pipelineMap = new LinkedHashMap<String, Object>();
+        pipelineMap.put("name", pipeline.getName());
+        pipelineMap.put("counter", new Integer(pipeline.getCounter()).toString());
+        pipelineMap.put("label", pipeline.getLabel());
+
+        List<Map> materialRevisionList = createBuildCauseDataMap(pipeline.getMaterialRevisions());
+        pipelineMap.put("build-cause", materialRevisionList);
+
+        List<Map> stageList = new ArrayList<Map>();
+        for (Stage currentStage : pipeline.getStages()) {
+            Map<String, Object> stageMap = createStageDataMap(currentStage);
+            stageList.add(stageMap);
+        }
+        pipelineMap.put("stages", stageList);
+
+        return pipelineMap;
+    }
+
+    private List<Map> createBuildCauseDataMap(MaterialRevisions materialRevisions) {
+        List<Map> materialRevisionList = new ArrayList<Map>();
+        for (MaterialRevision currentRevision : materialRevisions) {
+            Map<String, Object> materialRevisionMap = new LinkedHashMap<String, Object>();
+
+            Map<String, Object> materialMap = createMaterialDataMap(currentRevision.getMaterial());
+            materialRevisionMap.put("material", materialMap);
+            materialRevisionMap.put("changed", currentRevision.isChanged());
+
+            List<Map> modificationList = new ArrayList<Map>();
+            for (Modification modification : currentRevision.getModifications()) {
+                Map<String, Object> modificationMap = createModificationDataMap(modification);
+                modificationList.add(modificationMap);
+            }
+            materialRevisionMap.put("modifications", modificationList);
+
+            materialRevisionList.add(materialRevisionMap);
+        }
+        return materialRevisionList;
+    }
+
+    private Map<String, Object> createMaterialDataMap(Material material) {
+        Map<String, Object> materialMap = new LinkedHashMap<String, Object>();
+        if (material instanceof GitMaterial) {
```

Comment:
All of these should be in their own classes. That's why polymorphic dispatch exists. :) I realize they don't all share an interface. Make them. Something like, InformationProvider, and all kinds of materials implement that interface and they decide what information to put in their own maps.